investigate ws connection losses

after a while, the browser console shows many 

```
 ￼Bbp0PqUz.js:49 [fetchIncogniteeNotes] error Error: No response received from RPC endpoint in 60s
17
￼Cg5A4fbM.js:14 Error: No response received from RPC endpoint in 60s
    at ji.__internal__timeoutHandlers (￼CkpypJWy.js:100:19166)
    at ￼CkpypJWy.js:100:21381
￼bQho2aB0.js:1 2025-01-30 13:43:10          API-WS: disconnected from ￼wss://scv1.asset-hub-polkadot.api.incognitee.io:443: 1006:: Abnormal Closure
2
￼Bbp0PqUz.js:49 [fetchIncogniteeNotes] error Error: disconnected from ￼wss://scv1.asset-hub-polkadot.api.incognitee.io:443: 1006:: Abnormal Closure
2
￼Cg5A4fbM.js:14 Error: disconnected from ￼wss://scv1.asset-hub-polkadot.api.incognitee.io:443: 1006:: Abnormal Closure
    at WebSocket.__internal__onSocketClose (￼CkpypJWy.js:100:15983)
```

we also see errors on the worker side logs:

```
[2025-01-30T10:46:28.083721Z ERROR itc_tls_websocket_server::stream_state] Web-socket handshake failed: Protocol(MissingConnectionUpgradeHeader)
[2025-01-30T08:34:54.948795Z ERROR itc_tls_websocket_server::connection] cannot process TLS packet(s), closing connection: CorruptMessage
```

--- a/package.json
+++ b/package.json
@@ -44,8 +44,8 @@
     "vue-router": "^4.3.2"
   },
   "dependencies": {
-    "@encointer/util": "^0.18.1",
-    "@encointer/worker-api": "^0.18.1",
+    "@encointer/util": "^0.18.2",
+    "@encointer/worker-api": "^0.18.2",
     "@esbuild-plugins/node-globals-polyfill": "^0.2.3",
     "@headlessui/tailwindcss": "^0.2.1",
     "@headlessui/vue": "^1.7.22",

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -814,6 +814,13 @@ async function reconnectShieldingTargetIfNecessary() {
     );
     shieldingTargetApi.value = await ApiPromise.create({ provider: wsProvider });
     await shieldingTargetApi.value.isReady;
+
+    // await is quick as we only subscribe
+    await shieldingTargetApi.value.rpc.chain.subscribeNewHeads((lastHeader) => {
+      systemHealth.observeShieldingTargetBlockNumber(
+          lastHeader.number.toNumber(),
+      );
+    });
   }
 }
 

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -710,6 +710,11 @@ const pollingInterval = 2000; // 2 seconds
 let pollingTimeout: any = null;
 const isPolling = ref(true);
 
+// After 20 seconds in background, we will stop some services,
+// which can be resumed upon coming into foreground again.
+const inactivityTimeoutPeriod = 20000; // 20 seconds
+let disconnectWsTimeout: any = null;
+
 async function pollWorker() {
   if (!incogniteeStore.api?.isReady) {
     // Schedule the next polling.
@@ -737,20 +742,33 @@ function handleVisibilityChange() {
 
 async function onVisible() {
   isPolling.value = true;
+  clearInterval(disconnectWsTimeout);
+
   if (!incogniteeStore.api?.isConnected) {
     console.debug("[onVisible] Reconnecting to the worker api");
     await incogniteeStore.api?.connect();
   }
 
+  // avoid spamming polls due to visibility changes
+  clearInterval(pollingTimeout);
   await pollWorker();
 }
 
 async function onBackground() {
   isPolling.value = false;
+  clearTimeout(pollingTimeout);
+
   // otherwise we will see errors in the log that
   // the websocket unexpectedly closed after a while.
+
+  // prevent scheduling multiple closes
+  clearInterval(disconnectWsTimeout);
+  disconnectWsTimeout = setTimeout(closeWs, inactivityTimeoutPeriod);
+}
+
+async function closeWs() {
+  console.debug("closing websocket");
   await incogniteeStore.api?.closeWs();
-  clearTimeout(pollingTimeout);
 }
 
 watch(

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -676,20 +676,9 @@ const fetchIncogniteeNotes = async (
   }
 };
 
-const pollingInterval = 2000; // 2 seconds
-let pollingTimeout: any = null;
-const isPolling = ref(true);
-
-async function pollWorker() {
+async function fetchWorkerData() {
   if (!incogniteeStore.api?.isReady) {
-    // Schedule the next polling.
-    pollingTimeout = setTimeout(pollWorker, pollingInterval);
     return;
-  }
-
-  if (!incogniteeStore.api?.isConnected) {
-    // If we resumed from the background, it could be that we have to reconnect to the websocket again
-    await incogniteeStore.api?.connect();
   }
 
   console.debug(
@@ -715,6 +704,20 @@ async function pollWorker() {
   } catch (error) {
     console.warn("error auto-fetching older incognitee note buckets: " + error);
   }
+}
+
+const pollingInterval = 2000; // 2 seconds
+let pollingTimeout: any = null;
+const isPolling = ref(true);
+
+async function pollWorker() {
+  if (!incogniteeStore.api?.isReady) {
+    // Schedule the next polling.
+    pollingTimeout = setTimeout(pollWorker, pollingInterval);
+    return;
+  }
+
+  await fetchWorkerData();
 
   if (isPolling.value) {
     // Schedule the next polling.
@@ -724,14 +727,30 @@ async function pollWorker() {
 
 function handleVisibilityChange() {
   if (document.visibilityState === "visible") {
-    console.debug("[Polling] Tab is active. Resuming polling...");
-    isPolling.value = true;
-    pollWorker();
+    console.debug("[Polling] Tab became visible");
+    onVisible();
   } else {
-    console.debug("[Polling] Tab is inactive. Pausing polling...");
-    isPolling.value = false;
-    clearTimeout(pollingTimeout);
+    console.debug("[Polling] Tab became invisible");
+    onBackground();
   }
+}
+
+async function onVisible() {
+  isPolling.value = true;
+  if (!incogniteeStore.api?.isConnected) {
+    console.debug("[onVisible] Reconnecting to the worker api");
+    await incogniteeStore.api?.connect();
+  }
+
+  await pollWorker();
+}
+
+async function onBackground() {
+  isPolling.value = false;
+  // otherwise we will see errors in the log that
+  // the websocket unexpectedly closed after a while.
+  await incogniteeStore.api?.closeWs();
+  clearTimeout(pollingTimeout);
 }
 
 watch(

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -746,7 +746,7 @@ async function onVisible() {
 
   if (!incogniteeStore.api?.isConnected) {
     console.debug("[onVisible] Reconnecting to the worker api");
-    await incogniteeStore.api?.connect();
+    await incogniteeStore.api?.reconnect();
   }
 
   // avoid spamming polls due to visibility changes
@@ -768,7 +768,14 @@ async function onBackground() {
 
 async function closeWs() {
   console.debug("closing websocket");
-  await incogniteeStore.api?.closeWs();
+  if (incogniteeStore.api?.isConnected) {
+    // not sure why, but sometimes the websocket is already
+    // closed here.
+    await incogniteeStore.api?.closeWs();
+    console.debug("closed websocket");
+  } else {
+    console.debug("websocket was closed already");
+  }
 }
 
 watch(

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -676,8 +676,22 @@ const fetchIncogniteeNotes = async (
   }
 };
 
-const pollCounter = useInterval(2000);
-watch(pollCounter, async () => {
+const pollingInterval = 2000; // 2 seconds
+let pollingTimeout: any = null;
+const isPolling = ref(true);
+
+async function pollWorker() {
+  if (!incogniteeStore.api?.isReady) {
+    // Schedule the next polling.
+    pollingTimeout = setTimeout(pollWorker, pollingInterval);
+    return;
+  }
+
+  if (!incogniteeStore.api?.isConnected) {
+    // If we resumed from the background, it could be that we have to reconnect to the websocket again
+    await incogniteeStore.api?.connect();
+  }
+
   console.debug(
     `[IntegriteeWorker]: connections stats: ${JSON.stringify(incogniteeStore?.api?.wsStats)}`,
   );
@@ -701,7 +715,24 @@ watch(pollCounter, async () => {
   } catch (error) {
     console.warn("error auto-fetching older incognitee note buckets: " + error);
   }
-});
+
+  if (isPolling.value) {
+    // Schedule the next polling.
+    pollingTimeout = setTimeout(pollWorker, pollingInterval);
+  }
+}
+
+function handleVisibilityChange() {
+  if (document.visibilityState === "visible") {
+    console.debug("[Polling] Tab is active. Resuming polling...");
+    isPolling.value = true;
+    pollWorker();
+  } else {
+    console.debug("[Polling] Tab is inactive. Pausing polling...");
+    isPolling.value = false;
+    clearTimeout(pollingTimeout);
+  }
+}
 
 watch(
   () => accountStore.getAddress,
@@ -869,6 +900,7 @@ const switchToFaq = () => {
 onMounted(async () => {
   checkIfMobile();
   window.addEventListener("resize", checkIfMobile);
+  document.addEventListener("visibilitychange", handleVisibilityChange);
   await loadEnv(props.envFile);
   await incogniteeStore.initializeApi(
     chainConfigs[incogniteeSidechain.value].api,
@@ -921,11 +953,16 @@ onMounted(async () => {
     // if we move back from TEERdays, the account may already be selected and the subscription watcher won't trigger
     await subscribeWhatsReady();
   }
+  // start polling the worker
+  await pollWorker();
 });
 
 onUnmounted(() => {
   eventBus.off("addressClicked", openChooseWalletOverlay);
   window.removeEventListener("resize", checkIfMobile);
+
+  clearTimeout(pollingTimeout);
+  document.removeEventListener("visibilitychange", handleVisibilityChange);
 });
 
 const dropSubscriptions = () => {

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -218,6 +218,10 @@ const fetchIncogniteeBalance = async () => {
     return;
   }
 
+  if (!incogniteeStore.api?.isConnected) {
+    await incogniteeStore.api?.reconnect();
+  }
+
   isUpdatingIncogniteeBalance.value = true;
 
   const injector = accountStore.hasInjector ? accountStore.injector : null;
@@ -314,6 +318,9 @@ const fetchNetworkStatus = async () => {
     promises.push(p);
   }
   if (!incogniteeStore.apiReady) return;
+  if (!incogniteeStore.api?.isConnected) {
+    await incogniteeStore.api?.reconnect();
+  }
   console.debug("fetch network status info");
   const getter = incogniteeStore.api.parentchainsInfoGetter(
     incogniteeShard.value,
@@ -364,6 +371,10 @@ const updateNotes = async () => {
 };
 const fetchNoteBucketsInfo = async () => {
   if (!incogniteeStore.apiReady) return;
+  if (!incogniteeStore.api?.isConnected) {
+    await incogniteeStore.api?.reconnect();
+  }
+
   console.log("fetch note buckets info");
   const getter = incogniteeStore.api.noteBucketsInfoGetter(
     incogniteeStore.shard,
@@ -425,6 +436,11 @@ const fetchIncogniteeNotes = async (
     );
     return;
   }
+
+  if (!incogniteeStore.api?.isConnected) {
+    await incogniteeStore.api?.reconnect();
+  }
+
   const bucketIndex = maybeBucketIndex ? maybeBucketIndex : 0;
   const mapKey = `notesFor:${accountStore.account}:${bucketIndex}`;
   const sessionProxy = accountStore.sessionProxyForRole(
@@ -679,6 +695,10 @@ const fetchIncogniteeNotes = async (
 async function fetchWorkerData() {
   if (!incogniteeStore.api?.isReady) {
     return;
+  }
+
+  if (!incogniteeStore.api?.isConnected) {
+    await incogniteeStore.api?.reconnect();
   }
 
   console.debug(

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -810,15 +810,17 @@ async function reconnectShieldingTargetIfNecessary() {
   if (!shieldingTargetApi.value?.isConnected) {
     const wsProvider = new WsProvider(chainConfigs[shieldingTarget.value].api);
     console.log(
-        "re-initializing api at " + chainConfigs[shieldingTarget.value].api,
+      "re-initializing api at " + chainConfigs[shieldingTarget.value].api,
     );
-    shieldingTargetApi.value = await ApiPromise.create({ provider: wsProvider });
+    shieldingTargetApi.value = await ApiPromise.create({
+      provider: wsProvider,
+    });
     await shieldingTargetApi.value.isReady;
 
     // await is quick as we only subscribe
     await shieldingTargetApi.value.rpc.chain.subscribeNewHeads((lastHeader) => {
       systemHealth.observeShieldingTargetBlockNumber(
-          lastHeader.number.toNumber(),
+        lastHeader.number.toNumber(),
       );
     });
   }

--- a/pages/index.vue
+++ b/pages/index.vue
@@ -306,6 +306,9 @@ const storeSessionProxies = (proxies) => {
 const fetchNetworkStatus = async () => {
   const promises = [];
   if (shieldingTargetApi.value?.isReady) {
+    // If the browser tab was open, but inactive for a long time,
+    // it could be that we need to reconnect (ready is a bad indicator here).
+    await reconnectShieldingTargetIfNecessary();
     const p = shieldingTargetApi.value.rpc.chain
       .getFinalizedHead()
       .then((head) => {
@@ -802,6 +805,17 @@ watch(
   () => accountStore.getAddress,
   async () => await subscribeWhatsReady(),
 );
+
+async function reconnectShieldingTargetIfNecessary() {
+  if (!shieldingTargetApi.value?.isConnected) {
+    const wsProvider = new WsProvider(chainConfigs[shieldingTarget.value].api);
+    console.log(
+        "re-initializing api at " + chainConfigs[shieldingTarget.value].api,
+    );
+    shieldingTargetApi.value = await ApiPromise.create({ provider: wsProvider });
+    await shieldingTargetApi.value.isReady;
+  }
+}
 
 const subscribeWhatsReady = async () => {
   //todo! only reinitialize if account changes

